include/cachemere/cache.hpp

```diff
@@ -150,93 +167,123 @@ void Cache<K, V, I, E, SV, SK, TS>::retain(P predicate_fn)
     }
 }
 
-template<class K, class V, template<class, class> class I, template<class, class> class E, class SV, class SK, bool TS>
-void Cache<K, V, I, E, SV, SK, TS>::swap(CacheType& other)
+template<class K, class V, template<class, class> class I, template<class, class> class E, template<class, class> class C, class SV, class SK, bool TS>
+void Cache<K, V, I, E, C, SV, SK, TS>::swap(CacheType& other)
 {
     // Acquire both cache locks.
     std::unique_lock<std::recursive_mutex> me_guard(lock());
     std::unique_lock<std::recursive_mutex> other_guard(other.lock());
 
     using std::swap;
 
-    swap(m_current_size, other.m_current_size);
-    swap(m_maximum_size, other.m_maximum_size);
     swap(m_statistics_window_size, other.m_statistics_window_size);
 
     swap(m_insertion_policy, other.m_insertion_policy);
     swap(m_eviction_policy, other.m_eviction_policy);
-
-    swap(m_measure_key, other.m_measure_key);
-    swap(m_measure_value, other.m_measure_value);
+    swap(m_constraint_policy, other.m_constraint_policy);
 
     swap(m_data, other.m_data);
 
     swap(m_hit_rate_acc, other.m_hit_rate_acc);
     swap(m_byte_hit_rate_acc, other.m_byte_hit_rate_acc);
 }
 
-template<class K, class V, template<class, class> class I, template<class, class> class E, class SV, class SK, bool TS>
-inline size_t Cache<K, V, I, E, SV, SK, TS>::number_of_items() const
+template<class K, class V, template<class, class> class I, template<class, class> class E, template<class, class> class C, class SV, class SK, bool TS>
+inline size_t Cache<K, V, I, E, C, SV, SK, TS>::number_of_items() const
 {
     std::unique_lock<std::recursive_mutex> guard(lock());
     return m_data.size();
 }
 
-template<class K, class V, template<class, class> class I, template<class, class> class E, class SV, class SK, bool TS>
-inline size_t Cache<K, V, I, E, SV, SK, TS>::size() const
+template<class K, class V, template<class, class> class I, template<class, class> class E, template<class, class> class C, class SV, class SK, bool TS>
+template<typename... Args>
+void Cache<K, V, I, E, C, SV, SK, TS>::update_constraint(Args... args)
 {
     std::unique_lock<std::recursive_mutex> guard(lock());
-    return m_current_size;
+    m_constraint_policy->update(std::forward<Args>(args)...);
+
+    auto should_evict_another = [&](auto victim_it) { return victim_it != m_eviction_policy->victim_end() && !m_constraint_policy->is_satisfied(); };
+
+    for (auto victim_it = m_eviction_policy->victim_begin(); should_evict_another(victim_it); victim_it = m_eviction_policy->victim_begin()) {
```

Comment:
Just to clarify, you can update the _parameters_ of the constraint, but not the _constraint type_ right? Trying to wrap my head around modifying for example a TTL-based constraint. I kind of feel like the eviction policy would not necessarily be aware of whether its victims fit the bill in that case? This isn't really an issue with memory or number of element constraints since those don't need additional information about the elements, evicting any of them contributes to adhering to the constraint globally.

Reply:
Yep you can update the parameters only. I'll address the TTL comments in another reply since it's tricky (and could likely be improved)